labs/a4.md

```diff
@@ -0,0 +1,70 @@
+---
+title: Lab 4 - Post Install
+layout: lab
+---
+
+# Generating SSH Keys
+Since passwords can be guessed, a secure SSH setup will usually use key-based authentication instead of (or in addition to) a passphrase. You may already have an ssh key, so run the following command to see if you have one:
+
+(note: We won’t give instructions here for Windows machines. To follow these instructions, SSH into `ssh.ocf.berkeley.edu` with your OCF account’s credentials and continue from there)
+
+```
+ls .ssh
+```
+
+If you see id_rsa or id_edsca listed, you already have an SSH key. If you don’t, run this command to generate one:
+
+
+```
+ssh-keygen -t rsa -b 4096
+```
+
+Now, run `ls .ssh` again. You should see two files, `id_rsa` and `id_rsa.pub`.
+
+*  `id_rsa` is the private key-- it should never be revealed outside of the machine it’s generated on. Knowledge of this key can be used to enter your server!
+* `id_rsa.pub` is the public key. You can safely give it to your friends and/or shout it from the rooftops. If you put it in the `~/.ssh/authorized_keys` folder of a user on a server, you will be able to log in as that user on that server.
+
+Use this command to copy your public key to your decal VM, replacing <user> with your OCF username:
+
+
+```
+ssh-copy-id <user>@<user>.decal.xcf.sh
+```
+
+Now ssh into the decal VM (`ssh <user>@<user>.decal.xcf.sh`). If you set an SSH key passphrase, you should be prompted to enter it. If you did not set a passphrase, SSH should log you in immediately.
+
+
+To see what `ssh-copy-id` did: run `cat ~/.ssh/id_rsa.pub` on your local machine. Then run `cat ~/.ssh/authorized_keys` on your decal VM. You should see your public key listed in `authorized_keys`!
+
+Lab submission question: paste your `id_rsa.pub`
+
+# Install a firewall
+We have already set up a firewall outside your VM’s control that prevents connections from outside UC Berkeley’s subnets. However, as an educational exercise you should still install a firewall.
+
+Install UFW. You might find [this guide](https://wiki.debian.org/Uncomplicated%20Firewall%20%28ufw%29) helpful. Ensure the following
+* Incoming connections should be forbidden by default
+* Incoming connections to port 22 should be allowed (otherwise, you’ll be locked out of your VM!)
+* Incoming connections to ports 80 and 443 (HTTP and HTTPS) should be allowed.
```

Comment:
Really? Should they be allowed even when you have no plans to host a web server?

Reply:
I'll put a note that says this is optional.